onnxruntime/core/xnnpack/schema/xnnpack_onnx_defs.cc

```diff
@@ -0,0 +1,281 @@
+    return OnnxStatus(StatusCategory::NONE, StatusCode::FAIL, "Input tensor must have 4 dimensions");
+  }
+
+  if (weight_shape.dim_size() != 4) {
+    return OnnxStatus(StatusCategory::NONE, StatusCode::FAIL, "Weight tensor must have 4 dimensions");
+  }
+
+  int64_t input_H = input_shape.dim(1).dim_value();
+  int64_t input_W = input_shape.dim(2).dim_value();
+  int64_t input_C = input_shape.dim(3).dim_value();
+
+  if (input_C == 0) {
+    return OnnxStatus(StatusCategory::NONE, StatusCode::FAIL, "Input channel can not be zero");
+  }
+
+  // Weight shape: [1, kernel_height, kernel_width, input_channels * depth_multiplier]
+  int64_t size_one = weight_shape.dim(0).dim_value();
+  if (size_one != 1) {
+    std::ostringstream oss;
+    oss << "The first dim of weight must be one. Got " << size_one << ", " << input_H << ", " << input_W << ", "
+        << input_C << ".";
+    return OnnxStatus(StatusCategory::NONE, StatusCode::FAIL, oss.str());
+  }
+  int64_t filter_height = weight_shape.dim(1).dim_value();
+  int64_t filter_width = weight_shape.dim(2).dim_value();
+#if 0
+  int64_t input_channels_by_depth_multiplier = weight_shape.dim(3).dim_value();
+  if (input_channels_by_depth_multiplier % input_C != 0) {
+    return OnnxStatus(StatusCategory::NONE, StatusCode::FAIL,
+                      "The last dim of weight is not multiple of input channels.");
+  }
+#endif
+  input_H += static_cast<int64_t>(input_padding_top) + input_padding_bottom;
+  input_W += static_cast<int64_t>(input_padding_right) + input_padding_left;
+  ::ONNX_NAMESPACE::TensorShapeProto_Dimension* output_dims[4];
+  output_dims[0] = final_output_shape->add_dim();
+  output_dims[1] = final_output_shape->add_dim();
+  output_dims[2] = final_output_shape->add_dim();
+  output_dims[3] = final_output_shape->add_dim();
+  ONNX_RETURN_IF_ERROR(ConvShapeInference(input_shape.dim(0), input_H, input_W, input_C, weight_shape.dim(3),
+                                          filter_height, filter_width, input_C, subsampling_height, subsampling_width,
+                                          dilation_h, dilation_w, padding_mode, output_dims));
+  return OnnxStatus::OK();
+}
+
+// Compare to the signatures of xnn_define_convolution_2d function, this schema doesn't have
+// 1. kernel_height. Because it is just a dimension size of the weights
+// 2. kernel_width. Because it is just a dimension size of the weights
+// 3. group_input_channels. number of input channels per group. Can be calculated if input channels and the number of
+// groups are known
+// 4. group_output_channels. As the above
+ONNX_XNNPACK_OPERATOR_SET_SCHEMA(
+    XnnPackConvolution2d, 1,
+    OpSchema()
+        .Input(0, "X", "", "tensor(float)")
+        .Input(1, "W", "", "tensor(float)")
+        .Input(2, "B", "", "tensor(float)")
+        .Output(0, "X1", "", "tensor(float)")
+        .Attr("input_padding_top", "Implicit zero-padding above 2D input data. Must be 0 if padding mode is SAME",
```

Comment:
>Implicit [](http://example.com/codeflow?start=36&length=8)

Do you mean Explicit

Reply:
No. The comment was copied from xnnpack.